// @genkit-ai/google-cloud/lib/telemetry/defaults.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  TelemetryConfigs: () => TelemetryConfigs
});
module.exports = __toCommonJS(defaults_exports);
var import_core = require("@genkit-ai/core");
var import_sdk_trace_base = require("@opentelemetry/sdk-trace-base");
const TelemetryConfigs = {
  defaults: (overrides = {}) => {
    return (0, import_core.isDevEnv)() ? TelemetryConfigs.developmentDefaults(overrides) : TelemetryConfigs.productionDefaults(overrides);
  },
  developmentDefaults: (overrides = {}) => {
    const defaults = {
      sampler: new import_sdk_trace_base.AlwaysOnSampler(),
      autoInstrumentation: true,
      autoInstrumentationConfig: {
        "@opentelemetry/instrumentation-dns": { enabled: false }
      },
      instrumentations: [],
      metricExportIntervalMillis: 5e3,
      metricExportTimeoutMillis: 5e3,
      disableMetrics: false,
      disableTraces: false,
      exportIO: !overrides.disableLoggingIO,
      export: !!overrides.forceDevExport
      // false
    };
    return __spreadValues(__spreadValues({}, defaults), overrides);
  },
  productionDefaults: (overrides = {}) => {
    const defaults = {
      sampler: new import_sdk_trace_base.AlwaysOnSampler(),
      autoInstrumentation: true,
      autoInstrumentationConfig: {
        "@opentelemetry/instrumentation-dns": { enabled: false }
      },
      instrumentations: [],
      metricExportIntervalMillis: 3e5,
      metricExportTimeoutMillis: 3e5,
      disableMetrics: false,
      disableTraces: false,
      disableLoggingIO: false,
      exportIO: !overrides.disableLoggingIO,
      export: true
    };
    return __spreadValues(__spreadValues({}, defaults), overrides);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TelemetryConfigs
});
