// @genkit-ai/google-cloud/lib/telemetry/defaults.mjs
import {
  __spreadValues
} from "../chunk-DJRN6NKF.mjs";
import { isDevEnv } from "@genkit-ai/core";
import { AlwaysOnSampler } from "@opentelemetry/sdk-trace-base";
const TelemetryConfigs = {
  defaults: (overrides = {}) => {
    return isDevEnv() ? TelemetryConfigs.developmentDefaults(overrides) : TelemetryConfigs.productionDefaults(overrides);
  },
  developmentDefaults: (overrides = {}) => {
    const defaults = {
      sampler: new AlwaysOnSampler(),
      autoInstrumentation: true,
      autoInstrumentationConfig: {
        "@opentelemetry/instrumentation-dns": { enabled: false }
      },
      instrumentations: [],
      metricExportIntervalMillis: 5e3,
      metricExportTimeoutMillis: 5e3,
      disableMetrics: false,
      disableTraces: false,
      exportIO: !overrides.disableLoggingIO,
      export: !!overrides.forceDevExport
      // false
    };
    return __spreadValues(__spreadValues({}, defaults), overrides);
  },
  productionDefaults: (overrides = {}) => {
    const defaults = {
      sampler: new AlwaysOnSampler(),
      autoInstrumentation: true,
      autoInstrumentationConfig: {
        "@opentelemetry/instrumentation-dns": { enabled: false }
      },
      instrumentations: [],
      metricExportIntervalMillis: 3e5,
      metricExportTimeoutMillis: 3e5,
      disableMetrics: false,
      disableTraces: false,
      disableLoggingIO: false,
      exportIO: !overrides.disableLoggingIO,
      export: true
    };
    return __spreadValues(__spreadValues({}, defaults), overrides);
  }
};
export {
  TelemetryConfigs
};
